lib/blockchain.go

```diff
@@ -5747,3 +5747,126 @@ func (bc *Blockchain) CreateCoinUnlockTxn(
+// to subsidize all likes for users using atomic transactions. The following is
+// the step-by-step procedure by which the app can do so:
+//
+//	(1) The user creates an unsigned LIKE transaction as though they have DESO to pay for it.
+//		The LIKE transaction has no special metadata and is no different from an unsubsidized
+//		LIKE transaction up and to this point.
+//	(2) The user submits the unsigned LIKE transaction to the application to be subsidized
+//		as they cannot pay for the LIKE transaction themselves.
+//	(3) The application creates an unsigned BASIC_TRANSFER transaction to transfer the user enough
+//		DESO to cover their LIKE transaction. The transfer amount can be computed exactly
+//		using the TxnFeeNanos field of the provided LIKE transaction.
+//	(4) Rather than returning the BASIC_TRANSFER to the user, the application then calls
+//		CreateAtomicTxnsWrapper where unsignedTransactions = [BASIC_TRANSFER, LIKE]. This
+//		returns an atomic transaction that forces the user to use the BASIC_TRANSFER for this
+//		specific LIKE transaction. Note that this adds extra data to both the BASIC_TRANSFER
+//		and the LIKE transactions to ensure their atomic execution.
+//	(5) Before returning the atomic transaction to the user, the application now signs
+//		the BASIC_TRANSFER which resides within the atomic transaction wrapper on the server side.
+//		At this point, only the LIKE transaction remains unsigned within the atomic transaction.
+//	(6) The atomic transaction is returned to the user NOT the raw BASIC_TRANSFER transaction
+//		created in step 3.
+//	(7) The user signs the LIKE transaction using their private key on the client side.
+//		At this point, all transactions which reside within the AtomicTxns wrapper are signed.
+//	(8) The user submits the atomic transaction containing both the signed BASIC_TRANSFER
+//		transaction and the signed LIKE transaction to the blockchain.
+func (bc *Blockchain) CreateAtomicTxnsWrapper(
+	unsignedTransactions []*MsgDeSoTxn,
+	extraData map[string][]byte,
+) (
+	_txn *MsgDeSoTxn,
+	_fees uint64,
+	_err error,
+) {
+	// First we must convert the unsigned transactions into a doubly linked list via the
+	// transaction extra data. We create a copy of the transactions to ensure we do not
+	// modify the caller's data.
+
+	// Construct special helper functions for circular doubly linked list indexing.
+	nextIndex := func(currentIndex int, chainLength int) int {
+		// Check for the special case of an atomic chain of length 1.
+		if chainLength == 1 {
+			return currentIndex
+		}
+		return (currentIndex + 1) % chainLength
+	}
+	prevIndex := func(currentIndex int, chainLength int) int {
+		// Check for the special case of an atomic chain of length 1.
+		if chainLength == 1 {
+			return currentIndex
+		}
+
+		// Check for the wrap around case.
+		if currentIndex == 0 {
+			return chainLength - 1
+		}
+		return currentIndex - 1
+	}
```

Comment:
Update these based on my suggestions from [this link](https://go.dev/play/p/a7EbQAHqGoo)